Errors reporting

Mate2xo opened the pull request:
* integration  with Appsignal
* set distinct config for staging env
* improve system test perf
* promoted test helper to all system tests, as it's needed everywhere to use the headless mode

Mate2xo committed:
docs: update versions according to recent Rails update

Mate2xo committed:
feat(monitoring): Appsignal integration

Mate2xo committed:
test(system): headless JS browser, DRY helper method

System tests always spins up a JS browser, whether the :js tag is
present on examples/contexts or not. So we setup the driver manually,
only setting the JS environment when needed, to improve performance.

Mate2xo committed:
chore: lint

Mate2xo committed:
feat(monitoring): setup distinct staging config

Floury-Fabio commented on spec/support/helpers/system_browser.rb:
```diff
@@ -0,0 +1,18 @@
+    expect(page).to have_css('div#flash_messages')
+  end
+
+  def wait_flash(color)
```
suggestion: 
the name `wait_flash` doesn't seem follow the ecosystem rails conventions
We can rename this method wait_for_alert or expect_alert

We can also use wait_for_flash_messages for this method and remove the old method. The old method is not used. It was a fallback method that allowed the dev to wait for the flash without making any assumptions about its type

Mate2xo replied:
What conventions are you referring to ? You mean the default flash types ? (`alert` & `notice`)

Mate2xo replied:
Since the behaviour was previously exactly the same in these methods (except from the flash type)
```ruby
  def wait_for_success_flash
    expect(page).to have_css('div#flash_notice.alert-success')
  end

  def wait_for_warning_flash
    expect(page).to have_css('div#flash_alert.alert-warning')
  end

  def wait_for_danger_flash
    expect(page).to have_css('div#flash_error.alert-danger')
  end
```
I thought it was easier to read if we parameterized these types

Mate2xo replied:
Since they are actual expectations, I promoted them to be Matchers

Floury-Fabio commented on spec/system/document_uploads_at_document_page_section_spec.rb:
```diff
@@ -47,25 +49,24 @@
+
+      click_on I18n.t('main_app.views.application.buttons.destroy')
+      page.driver.browser.switch_to.alert.accept
     end
```
I agree with for the remove of the before, but  solution mixes several subjects and forward the problem of the before in the subjet.

my problems with this subject: 
-  the subject shouldn't contain the helper config `use_headless_javascript_browser`, this helper is not in the scenario from the point of view of the user.
- The subject merge the preparation and the exercice phase.
- The SUT is no longer explicit

proposition:

```
context 'context' do
    before { use_headless_javascript_browser }

    it 'it' do
      sign_in create(:member, :admin)
      create(:document)

      visit documents_path(anchor: 'documents')
      click_on I18n.t('main_app.views.application.buttons.destroy')
      page.driver.browser.switch_to.alert.accept

      expect(page).to have_no_content 'erd.pdf'
    end
  end
```
the before is still there but it is reserved for configuration and can be easily ignored in the context of the test.
The sut and the phases are explicit.
Repetition across tests is not a problem. Dry is counterproductive in testing.

The doc that guided me towards this proposal:

- https://thoughtbot.com/blog/the-self-contained-test
- https://thoughtbot.com/blog/how-we-test-rails-applications
- https://thoughtbot.com/blog/a-journey-towards-better-testing-practices
- https://github.com/thoughtbot/guides/tree/main/testing-rspec
- https://github.com/thoughtbot/guides/blob/main/testing-rspec/avoid_let_spec.rb

Floury-Fabio commented on README.md:
```diff
@@ -54,10 +54,17 @@ Members buy products from productors, and sell these products to themselves. Eac
+
+## Versions
+
+- Ruby => 2.7.8
```
you can change the versions :)

Mate2xo committed:
test: remove unused helper

Mate2xo committed:
test: promote 'expecting' to matcher

Mate2xo committed:
chore: lint

Mate2xo committed:
test: remove redundant test, improve subject representation

Mate2xo committed:
docs: remove useless info

Mate2xo committed:
test: remove now unnecessary wait simulation

Was previously needed, because on fast CPUs expectations on DB results
would be executed while requests were still pending

Floury-Fabio commented on spec/system/member_count_limit_on_missions_spec.rb:
```diff
@@ -1,66 +1,71 @@
-        click_button I18n.t('main_app.views.missions.show.button_enroll')
+    context 'when the enrolled Member count has been reached' do
+      let(:mission) do
+        create(:mission) do |mission|
+          mission.max_member_count = 4
+          mission.members << create_list(:member, 4)
+          mission.save
+        end
```
gray zone: you can avoid a save with factorybot params

Floury-Fabio commented on spec/system/members/cash_register_proficiency_spec.rb:
```diff
@@ -6,7 +6,10 @@
   let(:jack) { create(:member, cash_register_proficiency: :proficient) }
 
-  before { sign_in jack }
+  before do
```
gray zone: you shadow the driver config. you can instead make a config driver for each context putting the driver config in the before L15

Floury-Fabio commented on spec/system/members/cash_register_proficiency_spec.rb:
```diff
@@ -24,13 +27,14 @@
       visit missions_path
     end
 
     it 'shows missions without proficient members in purple', :js do
```
with driver config the js option is not necessary

Mate2xo committed:
test: simplify and explicit setup

Mate2xo committed:
test: scope config and resource creation to test context

Mate2xo committed:
test: the :js option is useless in system tests
